### PR TITLE
[Cpp][Qt][client] Fixed issue with unique items in OpenAPI schema

### LINKED ISSUE #11621: [BUG] qt-client missing qHash and operator== for schemes with unique models
#### Bug Report Checklist

- [x] How to reproduce ?
- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [x] Have you searched for related issues/PRs?
- [x] What's the actual output vs expected output?
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

##### Description

The generated code produced by the qt-client generator does not compile, if a scheme is included which includes lists of unique items.

##### openapi-generator version

* 5.2.0
* openapi-generator-cli-6.0.0-20211025.061654-22.jar (newest ?)

##### OpenAPI declaration file content or url

https://app.swaggerhub.com/apis/Pepperl-Fuchs/Sensorik4.0/1.0.0#/common.identification.identification_1

alternative:
www.pepperl-fuchs.com/openapi
-> click on Pepperl+Fuchs Sensorik4.0® will be redirected

The issue is rooted in the SCHEMAS:  
*  common.identification.identification
* common.identification.identification_1  

which include sets of unique objects.
##### Generation Details

Unsure what i should enter here: 
i am using vcpkg, c++, CMake and qt,  
platform: win10  
ide: vs2019

##### Steps to reproduce

0. Install dependencies

If you are using vcpkg:
My vcpkg.json 
```json
{
	"name": "openapisensorik40",
	"version-string" : "1.0",
	"description": "project to implement the openapisensorik40",
	"dependencies": [
		"qt5-base",
		"openssl"
	]
}
```

1. Download the mentioned oai specification
2. Run open-api-generator
3. Run Cmake to configure and build the library

An error is thrown, something like: 

qHash: none of the 37 overloads could convert argument types for 
OAICommon_identification_identification_1
OAICommon_identification_identification

##### Related issues/PRs

None?

##### Suggest a fix

From my point of view the issue is cause in some missing definition in order to make qhash(needed by QSet ) with a custom class work:  
Refer: https://doc.qt.io/qt-5/qhash.html#qhash    

I fixxed the issue by adding an operator== and a  hash function to the classes.

Wihtin OAICommon_identification_identification_1.h:
Add following function/ properties: to OAICommon_identification_identification_1
Within class definition (public)
```c++
bool operator==(const OAICommon_identification_identification_1& other) const {  
    return (this->asJsonObject() == other.asJsonObject());  
}  
```

End of file prior #endif

```c++
inline uint qHash(const OpenAPI::OAICommon_identification_identification_1& obj, uint seed = 0)
{
    return qHash(obj.asJsonObject(), seed);
}
```

Wihtin OAICommon_identification_identification_2.h:
Add following function/ properties: to OAICommon_identification_identification_2
Within class definition (public)
```c++
bool operator==(const OAICommon_identification_identification_2& other) const {  
    return (this->asJsonObject() == other.asJsonObject());  
}  
```
End of file prior #endif

```c++
inline uint qHash(const OpenAPI::OAICommon_identification_identification_2& obj, uint seed = 0)
{
    return qHash(obj.asJsonObject(), seed);
}
```

Well i think it is possible to add this to the corresponding mustache file, but right now i am pretty unsure where to start and if it is enough or need some additional coding. 

Can someone advice ?

### LINKED ISSUE #11951: [BUG] [Qt5][Cpp]  Unique items in schema lead to error
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [x] Have you searched for related issues/PRs?
- [x] What's the actual output vs expected output?
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

##### Description

If an array with unique items exists in the schema the generated code does not compile.
Because the data is stored in a QSet and output.append is caleld which is not present instead output.insert should be called

##### openapi-generator version

5.4.0.

##### OpenAPI declaration file content or url

Attached as file.

##### Generation Details

(Ubuntu 20.0)
java -jar openapi-generator-cli-5.4.0.jar generate -i unique_petstore.json -g cpp-qt-client -o unique_petstore

##### Steps to reproduce

Generate client via generator for cpp qt 
cmake -S . -B \build
cmake --build \build

Following error should appear: 

```
/home/thi/Projects/Github/oai-cpp-qt-client/unique_petstore/client/OAIPetsApi.cpp:388:16: error: ‘class QSet<OpenAPI::OAIPet>’ has no member named ‘append’  
  388 |         output.append(val);  
      |                ^~~~~~  
make[2]: *** [client/CMakeFiles/client.dir/build.make:105: client/CMakeFiles/client.dir/OAIPetsApi.cpp.o] Error 1  
make[2]: Leaving directory '/home/thi/Projects/Github/oai-cpp-qt-client/unique_petstore/build'  
make[1]: *** [CMakeFiles/Makefile2:98: client/CMakeFiles/client.dir/all] Error 2  
```

##### Suggest a fix

Coming soon

[unique_petstore.txt](https://github.com/OpenAPITools/openapi-generator/files/8334593/unique_petstore.txt)

### PR DESCRIPTION
**Pull request by Thaulino**
### Details of change 

object.mustache
Added equal operator 
Added qhash Operator

This way, the schema can be stored in a QSet
in api template
depending on unique items
output.insert(val) -- QSet (unique items)
or
ouput.append(val) -- QList (not unique items)

### Test done 
- [x] does the latest openapi-generator generate valid code  ?
I used the same openapi as attached to this issue  #11951
java -jar openapi-generator-cli-5.4.0.jar generate -t cpp-qt-templates -i unique_petstore.json -g cpp-qt-client -o unique_petstore
Build the library with cmake

### related issues
resolves #11621
resolves #11951

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (5.3.0), `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.
- [ ] 
### technical committee
@muttleyxd
@etherealjoy
@wing328

### CODE REVIEW

**Comment by wing328** [comment]
cc @ravinikam (2017/07) @stkrwork (2017/07) @etherealjoy (2018/02) @MartinDelille (2018/03) @muttleyxd (2019/08)

**Comment by canadaduane** [comment]
I think you mis-tagged me.

**Comment by Thaulino** [comment]
Okay now i have added a schema for the petstore with unique items plus a config to run the generator on that. 
I see some testcode within cpp-qt/PetStore which seem to run against a server. 
Unfortunately i don't know how all this is setup and tear down can someone advice ? 
Even if it is not needed, i would like to understand how it works

**Comment by wing328** [comment]
@canadaduane sorry

File: bin/configs/cpp-qt-client-petstore-unique.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+generatorName: cpp-qt-client
+outputDir: samples/client/petstore/cpp-qt-unqiue-items
+inputSpec: modules/openapi-generator/src/test/resources/2_0/petstore_plus_unqiue.json
```
Comment (MartinDelille) [inline bin/configs/cpp-qt-client-petstore-unique.yaml:3]:
```suggestion
inputSpec: modules/openapi-generator/src/test/resources/2_0/petstore_plus_unique.json
```

File: bin/configs/cpp-qt-client-petstore-unique.yaml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+generatorName: cpp-qt-client
+outputDir: samples/client/petstore/cpp-qt-unqiue-items
```
Comment (MartinDelille) [inline bin/configs/cpp-qt-client-petstore-unique.yaml:2]:
```suggestion
outputDir: samples/client/petstore/cpp-qt-unique-items
```

**Review by MartinDelille — CHANGES_REQUESTED** [review]
Little typo: unqiue => unique

**Comment by wing328** [comment]
Can you please add the file `modules/openapi-generator/src/test/resources/2_0/petstore_plus_unique.json` to the PR?

File: modules/openapi-generator/src/test/resources/2_0/petstore_plus_unique.json [hunk 3]
```diff
@@ -0,0 +1,206 @@
+{
+  "openapi": "3.0.0",
```
Comment (wing328) [inline modules/openapi-generator/src/test/resources/2_0/petstore_plus_unique.json:2]: it's a 3.0 spec. can you please put it under the 3_0 folder instead?

Comment (Thaulino) [reply modules/openapi-generator/src/test/resources/2_0/petstore_plus_unique.json]: Fixed it , Thx

**Comment by Thaulino** [comment]
Somehow Node 0 failed &#9785;
https://app.circleci.com/pipelines/github/OpenAPITools/openapi-generator/16507/workflows/f2b5272d-e91c-4f0e-b272-bea97a0ac288/jobs/34844?invite=true#step-114-871
Unfortunately i don't understand what is meant by the error output,
Need help/ Please help, thank you

**Comment by wing328** [comment]
For circleci failure, please ignore that for the time being.

**Comment by Thaulino** [comment]
@wing328 Is there anything left to do for me ?

**Comment by wing328** [comment]
@Thaulino thanks for the PR. I've merged https://github.com/OpenAPITools/openapi-generator/pull/12124 into master to use a different spec for C++ qt 5 petstore client.

Can you please rebase/merge latest master into your branch and add those tests in modules/openapi-generator/src/test/resources/3_0/petstore_plus_unique.json to modules/openapi-generator/src/test/resources/3_0/cpp-qt/petstore.yaml instead?

**Comment by wing328** [comment]
For CircleCI node 0 failure, please ignore it as it's not related to this change.

**Comment by Thaulino** [comment]
@wing328 I had some spare time and rebased like you requested. 
I deleted the obsolete stuff too (refer to https://github.com/OpenAPITools/openapi-generator/pull/11954/commits/79053b029783424eacaa4d7ff48aee07481fd642)
I updated the samples too. 
Please giv eme some feedback if it is fine or not - thank you

**Comment by wing328** [comment]
I tested it in the CI but got the following errors:
```
make -f client/CMakeFiles/client.dir/build.make client/CMakeFiles/client.dir/depend
make[2]: Entering directory '/home/travis/build/OpenAPITools/openapi-generator/samples/client/petstore/cpp-qt/build'
cd /home/travis/build/OpenAPITools/openapi-generator/samples/client/petstore/cpp-qt/build && /usr/local/cmake-3.12.4/bin/cmake -E cmake_depends "Unix Makefiles" /home/travis/build/OpenAPITools/openapi-generator/samples/client/petstore/cpp-qt /home/travis/build/OpenAPITools/openapi-generator/samples/client/petstore/cpp-qt/client /home/travis/build/OpenAPITools/openapi-generator/samples/client/petstore/cpp-qt/build /home/travis/build/OpenAPITools/openapi-generator/samples/client/petstore/cpp-qt/build/client /home/travis/build/OpenAPITools/openapi-generator/samples/client/petstore/cpp-qt/build/client/CMakeFiles/client.dir/DependInfo.cmake --color=
Scanning dependencies of target client
make[2]: Leaving directory '/home/travis/build/OpenAPITools/openapi-generator/samples/client/petstore/cpp-qt/build'
make -f client/CMakeFiles/client.dir/build.make client/CMakeFiles/client.dir/build
make[2]: Entering directory '/home/travis/build/OpenAPITools/openapi-generator/samples/client/petstore/cpp-qt/build'
[  8%] Building CXX object client/CMakeFiles/client.dir/PFXApiResponse.cpp.o
cd /home/travis/build/OpenAPITools/openapi-generator/samples/client/petstore/cpp-qt/build/client && /usr/bin/g++-5  -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_NO_DEBUG -I/home/travis/build/OpenAPITools/openapi-generator/samples/client/petstore/cpp-qt/build/client -I/home/travis/build/OpenAPITools/openapi-generator/samples/client/petstore/cpp-qt/client -I/home/travis/build/OpenAPITools/openapi-generator/samples/client/petstore/cpp-qt/build/client/client_autogen/include -isystem /usr/include/x86_64-linux-gnu/qt5 -isystem /usr/include/x86_64-linux-gnu/qt5/QtCore -isystem /usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++-64 -isystem /usr/include/x86_64-linux-gnu/qt5/QtNetwork -isystem /usr/include/x86_64-linux-gnu/qt5/QtGui  -fPIC -Wall -Wno-unused-variable -fPIC -Wall -Wno-unused-variable   -fPIC -std=c++14 -o CMakeFiles/client.dir/PFXApiResponse.cpp.o -c /home/travis/build/OpenAPITools/openapi-generator/samples/client/petstore/cpp-qt/client/PFXApiResponse.cpp
In file included from /home/travis/build/OpenAPITools/openapi-generator/samples/client/petstore/cpp-qt/client/PFXApiResponse.h:26:0,
                 from /home/travis/build/OpenAPITools/openapi-generator/samples/client/petstore/cpp-qt/client/PFXApiResponse.cpp:12:
/home/travis/build/OpenAPITools/openapi-generator/samples/client/petstore/cpp-qt/client/PFXObject.h: In function ‘uint test_namespace::qHash(const test_namespace::PFXObject&, uint)’:
/home/travis/build/OpenAPITools/openapi-generator/samples/client/petstore/cpp-qt/client/PFXObject.h:66:34: error: invalid initialization of reference of type ‘const test_namespace::PFXObject&’ from expression of type ‘QJsonObject’
     return qHash(obj.asJsonObject(), seed);
                                  ^
/home/travis/build/OpenAPITools/openapi-generator/samples/client/petstore/cpp-qt/client/PFXObject.h:65:13: note: in passing argument 1 of ‘uint test_namespace::qHash(const test_namespace::PFXObject&, uint)’
 inline uint qHash(const PFXObject& obj, uint seed = 0) noexcept{
             ^
client/CMakeFiles/client.dir/build.make:65: recipe for target 'client/CMakeFiles/client.dir/PFXApiResponse.cpp.o' failed
make[2]: Leaving directory '/home/travis/build/OpenAPITools/openapi-generator/samples/client/petstore/cpp-qt/build'
make[2]: *** [client/CMakeFiles/client.dir/PFXApiResponse.cpp.o] Error 1
CMakeFiles/Makefile2:164: recipe for target 'client/CMakeFiles/client.dir/all' failed
make[1]: Leaving directory '/home/travis/build/OpenAPITools/openapi-generator/samples/client/petstore/cpp-qt/build'
make[1]: *** [client/CMakeFiles/client.dir/all] Error 2
make: *** [all] Error 2
Makefile:132: recipe for target 'all' failed
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.2.1:exec (pet-test) on project CppQt5PetstoreClientTests: Command execution failed. Process exited with an error: 2 (Exit value: 2) -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.2.1:exec (pet-test) on project CppQt5PetstoreClientTests: Command execution failed.
```

Can you please take a look when you've time?

Ref: https://app.travis-ci.com/github/OpenAPITools/openapi-generator/builds/250142967

**Comment by Thaulino** [comment]
Hi, 

My interpretation of the error: Somehow the compiler doesn't find an overloaded version of qhash with the signature
qhash(QJsonObject, uint) so it tries to implicit cast the types which does not work. 

Disclaimer: I am not that deep into travis ci, containerization and so on

Educated guess:
As far as i understand the configuration, it is running on ubuntu xenial with g++-5, it would be interesting to know which qt version is running.
I guess this one: https://launchpad.net/ubuntu/xenial/+package/qtbase5-dev 
so version qt5.5, which is pretty old
Maybe this version had not implemented a qhash function for QJsonObjects. 

Next steps: 
i will try to run a docker container with the xenial image, install the qt base stuff and try to compile it. 
It it fails, == the qt version really does not implement a qhash function, the ubuntu version should be updated (travis ci)

it does not fail: my educated guess is wrong

**Comment by Thaulino** [comment]
Hi, 

I have created some minimal test setup based on docker
 https://github.com/Thaulino/qt5-ubuntu-docker-qhash-qjsonobj   
Well the build and passes on focal and fails on xenial with 
```
/usr/include/x86_64-linux-gnu/qt5/QtCore/qhash.h:104:16: note:   'const QJsonObject' is not derived from 'const QSharedDataPointer<T>'
 { return (qHash(t) ^ seed); }
                ^
/usr/include/x86_64-linux-gnu/qt5/QtCore/qhash.h:144:49: note: candidate: template<class T1, class T2> uint qHash(const QPair<T1, T2>&, uint)
 template <typename T1, typename T2> inline uint qHash(const QPair<T1, T2> &key,
                                                 ^
/usr/include/x86_64-linux-gnu/qt5/QtCore/qhash.h:144:49: note:   template argument deduction/substitution failed:
/usr/include/x86_64-linux-gnu/qt5/QtCore/qhash.h:104:16: note:   'const QJsonObject' is not derived from 'const QPair<T1, T2>'
 { return (qHash(t) ^ seed); }
```
Looks similiar to the issue in travis ci, so it seems like the needed qhash function isn't implemented  
```
root@d50123fb3380:/home# qmake -v
QMake version 3.0
Using Qt version 5.5.1 in /usr/lib/x86_64-linux-gnu
```

My suggestion is to update the ubuntu version used in travis ci
@wing328 : is that possible ? do you agree with the analysis ?

**Comment by wing328** [comment]
Tested locally with a Linux box instead and the result is good:
```
Totals: 5 passed, 0 failed, 0 skipped, 0 blacklisted, 29ms
********* Finished testing of StoreApiTests *********
********* Start testing of UserApiTests *********
Config: Using QtTest library 5.12.8, Qt 5.12.8 (x86_64-little_endian-lp64 shared (dynamic) release build; by GCC 9.3.0)
PASS   : UserApiTests::initTestCase()
QDEBUG : UserApiTests::createUserTest() got a request body
PASS   : UserApiTests::createUserTest()
QDEBUG : UserApiTests::createUsersWithArrayInputTest() got a request body
PASS   : UserApiTests::createUsersWithArrayInputTest()
QDEBUG : UserApiTests::createUsersWithListInputTest() got a request body
PASS   : UserApiTests::createUsersWithListInputTest()
QDEBUG : UserApiTests::deleteUserTest() Error happened while issuing request :  "Error transferring http://petstore.swagger.io/v2/user/rambo - server replied: Not Found, "
PASS   : UserApiTests::deleteUserTest()
QDEBUG : UserApiTests::getUserByNameTest() "johndoe"
PASS   : UserApiTests::getUserByNameTest()
QDEBUG : UserApiTests::loginUserTest() "{\"code\":200,\"type\":\"unknown\",\"message\":\"logged in user session:1651994014470\"}"
PASS   : UserApiTests::loginUserTest()
PASS   : UserApiTests::logoutUserTest()
QDEBUG : UserApiTests::updateUserTest() got a request body
PASS   : UserApiTests::updateUserTest()
PASS   : UserApiTests::cleanupTestCase()
Totals: 10 passed, 0 failed, 0 skipped, 0 blacklisted, 81ms
********* Finished testing of UserApiTests *********
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:36 min
[INFO] Finished at: 2022-05-08T15:13:34+08:00
[INFO] ------------------------------------------------------------------------
```

We will test locally instead of using Travis for the time being.

Thanks for the PR.

**Pull request merged**